### PR TITLE
Split auth related config for Azure

### PR DESCRIPTION
**Pull request by karataliu**
**What this PR does / why we need it**:
Azure credential provider should not depend on cloud provider.

This PR splits Azure config into auth related config and remaining. The auth related config could be reused by credential provider.
Currently the credential provider depends on cloud provider directly, after this change it only depends on Azure SDK.

**Which issue(s) this PR fixes**:
Part of Issue #50752 'Cleanup for Azure credential provider'

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/credentialprovider/azure/azure_auth.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	"io/ioutil"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/golang/glog"
+	"golang.org/x/crypto/pkcs12"
+)
+
+// AuthConfig holds auth related part of cloud config
+type AuthConfig struct {
```
Comment (brendandburns) [inline pkg/credentialprovider/azure/azure_auth.go]: its too bad you are duplicating this.is there a way we could share this definition somehow?

File: pkg/credentialprovider/azure/azure_auth.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+			config.AADClientID,
+			certificate,
+			privateKey,
+			env.ServiceManagementEndpoint)
+	}
+
+	return nil, fmt.Errorf("No credentials provided for AAD application %s", config.AADClientID)
+}
+
+func parseAzureEnvironment(cloudName string) (*azure.Environment, error) {
```
Comment (brendandburns) [inline pkg/credentialprovider/azure/azure_auth.go]: this code too. i am really alergic to this code duplication.

**Comment by karataliu** [comment]
@brendandburns Yeah, such duplication is painful.

This PR intends to remove `pkg/cloudprovider/providers/azure` from `pkg/credentialprovider/azure`'s dependency graph.

Currently we support 3 kinds of auth methods: certification, service principal, msi. And both `pkg/credentialprovider/azure` and other parts of `pkg/cloudprovider/providers/azure` now shares same logic for get auth token for arm. That's what I've extracted into 'azure_auth.go'

In order to minimize duplication, I could think about following:
1. Leave a single copy of 'azure_auth.go' in `pkg/cloudprovider/providers/azure`, and let `pkg/credentialprovider/azure` still depend on `pkg/cloudprovider/providers/azure`.
And copy it to `pkg/credentialprovider/azure` when we remove cloud provider code
2. Place `azure_auth.go` somewhere outside of both directories
3. Pending on out-of-tree credential provider, part of discussions could be found here: https://github.com/kubernetes/kubernetes/issues/48690

Suppose we could go with #1st first and pending on other solutions? Please kindly suggest if you have other thoughts.

**Comment by brendandburns** [comment]
Maybe we can add a new directory:

pkg/cloudprovider/providers/azure/auth

And have both depend on that?

That way we only have one copy, but we break the spaghetti dependencies somewhat.

--brendan

**Comment by karataliu** [comment]
@brendandburns @andyzhangx  Updated according to comments. PTAL

**Comment by karataliu** [comment]
golint is suggesting:
```
pkg/cloudprovider/providers/azure/auth/azure_auth.go:32:6: type name will be used as auth.AuthConfig by other packages, and that stutters; consider calling this Config
```
There's already a 'Config' struct type (for cloud config) under azure dir, I've updated the pr by renaming 'AuthConfig' to 'AzureAuthConfig'. PTAL

**Pull request merged**